mart/attack/adversary.py

```diff
@@ -275,14 +264,24 @@ def forward(
         model: Optional[torch.nn.Module] = None,
         **kwargs
     ):
+        # Backward compatible: earlier configs have two unnamed arguments [input, target].
+        kwargs["input"] = input
+        kwargs["target"] = target
+        kwargs["model"] = model
+
+        # The benign input to an adversary is not necessarily the model input.
+        # It could be activations of other modules.
+        self.benign_input = kwargs[self.benign_input_key]
+
         # Generate a perturbation only if we have a model. This will update
         # the parameters of self.perturber.
         if model is not None:
-            super().forward(input, target, model, **kwargs)
+            super().forward(**kwargs)
 
         # Get perturbation and apply threat model
-        perturbation = self.perturber(input, target)
-        output = self.threat_model(input, target, perturbation, **kwargs)
+        # The mask projector in perturber may require information from target.
+        perturbation = self.perturber(self.benign_input, target)
+        output = self.threat_model(self.benign_input, target, perturbation)
```

Comment:
I'm hoping this goes away but we shall see. If you can think of a way to rewrite any of the above so that this change does not need to be made, then that is a better solution because you don't break the abstraction. A one off solution like `self.benign_input` is a smell.